code/modules/hydroponics/biogenerator.dm

```diff
@@ -1,337 +1,354 @@
+	/// A list of plants currently stored plants in the biogenerator.
+	var/list/stored_plants = list()
+	/// The maximum amount of plants the biogenerator can store.
+	var/max_storable_plants = BASE_MAX_STORABLE_PLANTS
+	/// A reference to the biogenerator's research, which contains designs that it can build.
 	var/datum/research/files
-	var/list/show_categories = list("Food", "Botany Chemicals", "Organic Materials", "Leather and Cloth")
-	var/list/timesFiveCategories = list("Food", "Botany Chemicals", "Organic Materials")
+	/// A list which holds all categories and items the biogenator has available. Used with the UI to save having to rebuild this every time someone opens it.
+	var/list/product_list = list()
+	/// The categories in which the various items belong to.
+	var/list/categories = list("Food", "Botany Chemicals", "Organic Materials", "Leather and Cloth")
 
 /obj/machinery/biogenerator/Initialize(mapload)
 	. = ..()
-	files = new /datum/research/biogenerator(src)
-	create_reagents(1000)
+	files = new(src)
 	component_parts = list()
 	component_parts += new /obj/item/circuitboard/biogenerator(null)
 	component_parts += new /obj/item/stock_parts/matter_bin(null)
 	component_parts += new /obj/item/stock_parts/manipulator(null)
 	component_parts += new /obj/item/stack/sheet/glass(null)
 	component_parts += new /obj/item/stack/cable_coil(null, 1)
 	RefreshParts()
+	update_ui_product_list()
 
 /obj/machinery/biogenerator/Destroy()
-	QDEL_NULL(beaker)
+	QDEL_NULL(container)
 	QDEL_NULL(files)
+	QDEL_LIST(stored_plants)
 	return ..()
 
 /obj/machinery/biogenerator/ex_act(severity)
-	if(beaker)
-		beaker.ex_act(severity)
+	container?.ex_act(severity)
 	..()
 
 /obj/machinery/biogenerator/handle_atom_del(atom/A)
 	..()
-	if(A == beaker)
-		beaker = null
-		update_icon()
-		updateUsrDialog()
+	if(A != container)
+		return
+	container = null
+	update_icon()
+	SStgui.update_uis(src)
 
 /obj/machinery/biogenerator/RefreshParts()
-	var/E = 0
-	var/P = 0
-	var/max_storage = 40
+	efficiency = 0
+	productivity = 0
 	for(var/obj/item/stock_parts/matter_bin/B in component_parts)
-		P += B.rating
-		max_storage = 40 * B.rating
+		productivity += B.rating
+		max_storable_plants = BASE_MAX_STORABLE_PLANTS * B.rating
 	for(var/obj/item/stock_parts/manipulator/M in component_parts)
-		E += M.rating
-	efficiency = E
-	productivity = P
-	max_items = max_storage
-
-/obj/machinery/biogenerator/on_reagent_change()			//When the reagents change, change the icon as well.
-	update_icon()
+		efficiency += M.rating
 
 /obj/machinery/biogenerator/update_icon()
 	if(panel_open)
 		icon_state = "biogen-empty-o"
-	else if(!beaker)
+	else if(!container)
 		icon_state = "biogen-empty"
 	else if(!processing)
 		icon_state = "biogen-stand"
 	else
 		icon_state = "biogen-work"
-	return
+
+/obj/machinery/biogenerator/screwdriver_act(mob/living/user, obj/item/I)
+	if(!default_deconstruction_screwdriver(user, "biogen-empty-o", "biogen-empty", I) || !container)
+		return FALSE
```

Comment:
too many boolean ops in a single `if` for my liking. Consider
```suggestion
	if(!default_deconstruction_screwdriver(user, "biogen-empty-o", "biogen-empty", I))
		return FALSE
	if(!container)
		return FALSE
```

Separately, i think this should return TRUE even if no container is present - you can successfully open and close panel regardless of whether the container exists; it is only moved outside as a sort of convenience *in addition* to the main screwdriver operation.
```dm
	if(!default_deconstruction_screwdriver(user, "biogen-empty-o", "biogen-empty", I))
		return FALSE
	if(container)
		container.forceMove(loc)
		container = null
	return TRUE
```